### PR TITLE
cdc: support batch service

### PR DESCRIPTION
**Pull request by 5kbpers**
###  What have you changed?

Support to subscribe multiple regions by one stream in CDC.
Blocked by https://github.com/pingcap/kvproto/pull/567

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?

- Unit test
- Integration test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

Release notes.

###  Does this PR affect `tidb-ansible`?

No.

### CODE REVIEW

File: components/cdc/src/endpoint.rs [hunk 1]
```diff
@@ -146,38 +159,85 @@ impl<T: CasualRouter<RocksEngine>> Endpoint<T> {
...
-    pub fn on_register(&mut self, mut request: ChangeDataRequest, downstream: Downstream) {
+    pub fn on_register(
+        &mut self,
+        mut request: ChangeDataRequest,
+        mut downstream: Downstream,
+        conn_id: ConnID,
+    ) {
+        let conn = match self.connections.get_mut(&conn_id) {
+            Some(conn) => conn,
+            None => return,
```
[the remark this answers is not in the document]
Comment (5kbpers) [reply components/cdc/src/endpoint.rs]: In a subtle situation, the deregister cmd for deregistering the connection is sent before a register cmd.
In more cases, it must be an error, surely should print a log here.

File: components/cdc/src/endpoint.rs [hunk 2]
```diff
@@ -146,38 +159,85 @@ impl<T: CasualRouter<RocksEngine>> Endpoint<T> {
...
+                    // Do not continue to observe the events of the region
+                    self.observer.unsubscribe_region(region_id);
+                }
             }
-            (None, None) => panic!("none id none error?"),
-        }
-        if is_last {
-            // Unsubscribe region role change events.
-            self.observer.unsubscribe_region(region_id);
+            (None, None, Some(conn_id)) => {
```
[the remark this answers is not in the document]
Comment (5kbpers) [reply components/cdc/src/endpoint.rs]: That means a connection is closed.

File: components/cdc/src/delegate.rs [hunk 3]
```diff
@@ -549,23 +599,12 @@ fn decode_default(value: Vec<u8>, row: &mut EventRow) {
...
     use std::cell::Cell;
-    use std::sync::Arc;
-    use tikv::server::RaftKv;
     use tikv::storage::mvcc::test_util::*;
-    use tikv::storage::mvcc::tests::*;
-    use tikv_util::mpsc::{bounded, Sender as UtilSender};
+    use tikv_util::mpsc::batch::{self, SizedBatchReceiver};
 
     // TODO add test_txn once cdc observer is ready.
     // https://github.com/overvenus/tikv/blob/447d10ae80b5b7fc58a4bef4631874a11237fdcf/components/cdc/src/delegate.rs#L615-L701
```
Comment (overvenus) [inline components/cdc/src/delegate.rs]: Can we remove it?

File: components/cdc/src/delegate.rs [hunk 4]
```diff
@@ -25,14 +28,16 @@ use resolved_ts::Resolver;
...
 
 static DOWNSTREAM_ID_ALLOC: AtomicUsize = AtomicUsize::new(0);
+const EVENT_MAX_SIZE: usize = 6 * 1024 * 1024; // 6MB
 
 /// A unique identifier of a Downstream.
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
 pub struct DownstreamID(usize);
 
 impl DownstreamID {
```
Comment (overvenus) [inline components/cdc/src/delegate.rs]: Could you change SeqCst to Relaxed by the way?

File: components/cdc/src/delegate.rs [hunk 5]
```diff
@@ -45,36 +50,57 @@ impl DownstreamID {
...
+
+    pub fn sink_duplicate_error(&self, region_id: u64) {
+        let mut change_data_event = Event::default();
+        let mut cdc_err = EventError::default();
+        let mut err = ErrorDuplicateRequest::default();
+        err.set_region_id(region_id);
+        cdc_err.set_duplicate_request(err);
+        change_data_event.event = Some(Event_oneof_event::Error(cdc_err));
+        change_data_event.region_id = region_id;
+        self.sink_event(change_data_event, 0);
```
Comment (overvenus) [inline components/cdc/src/delegate.rs]: What does 0 mean? Could you add some comments?

File: components/tikv_util/src/mpsc/batch.rs [hunk 6]
```diff
@@ -298,6 +298,69 @@ impl<T, E, I: Fn() -> E, C: Fn(&mut E, T)> Stream for BatchReceiver<T, E, I, C>
     }
 }
 
+/// `SizedBatchReceiver` is a `futures::Stream`, which returns a batched type.
+/// It receives the size of the sent item, uses this size to compute the batch size.
+pub struct SizedBatchReceiver<T, E, I: Fn() -> E, C: Fn(&mut E, T)> {
```
Comment (overvenus) [inline components/tikv_util/src/mpsc/batch.rs]: Can it be replaced by [`Stream::chunks`](https://docs.rs/futures/0.1.29/futures/stream/trait.Stream.html#method.chunks)?

Comment (5kbpers) [reply components/tikv_util/src/mpsc/batch.rs]: Have replaced it with `flattenMap`.

File: components/cdc/src/delegate.rs (hunk 5 above)
Comment (5kbpers) [reply components/cdc/src/delegate.rs]: The size of error and resolved events is considered zero, I think it's easy to understand so not necessary to add comments for it (It appears in many places).

Comment (5kbpers) [reply components/cdc/src/delegate.rs]: Added the comment of `sink_event`

File: components/cdc/src/service.rs [hunk 7]
```diff
@@ -142,21 +143,35 @@ impl ChangeData for Service {
...
-            }
-            Err(e) => {
-                error!("cdc send failed"; "error" => ?e);
-                Err(Error::RpcFailure(RpcStatus::new(
-                    RpcStatusCode::UNKNOWN,
+        let rx = BatchReceiver::new(rx, CDC_MSG_MAX_BATCH_SIZE, Vec::new, |v, e| v.push(e));
+        let send_resp = sink.send_all(
+            rx.map(|events| {
+                let events_len = events.len();
+                let mut event_vecs = vec![Vec::with_capacity(events_len)];
```
Comment (overvenus) [inline components/cdc/src/service.rs]:
```suggestion
                let mut event_vecs = Vec::with_capacity(events_len);
```

Can we remove `stream::iter_ok` and `flatten`?

**Pull request merged**